MyRangesImplementation/MyRangesImplementation/cartezian_product_iterator.hpp

```diff
@@ -0,0 +1,112 @@
+					advanceCartezianIterator(std::make_index_sequence<
+						std::tuple_size_v<decltype(this->m_beginCurrenEndIterTuples)>
+					>{});
+					return *this;
+				}
+
+				cartezian_product_iterator<Iterators...> operator++(int) {
+					auto formerValue = *this;
+					++(*this);
+					return formerValue;
+				}
+
+				bool operator==(const cartezian_product_iterator<Iterators...>& other) const {
+					return this->m_beginCurrenEndIterTuples == other.m_beginCurrenEndIterTuples;
+				}
+
+				bool operator!=(const cartezian_product_iterator<Iterators...>& other) const {
+					return !(*this == other);
+				}
+
+				decltype(auto) operator*() const {
+					return makeTupleOfIteratorValues(std::make_index_sequence<
+						std::tuple_size_v<decltype(this->m_beginCurrenEndIterTuples)>>{});
+				}
+
+				auto getEndIterator() const {
+					auto copyOfThis = *this;
+					copyOfThis.setAllCurrentIteratorsToTheirEnd(
+						std::make_index_sequence<std::tuple_size_v<decltype(copyOfThis.m_beginCurrenEndIterTuples)>>{});
+					return copyOfThis;
+				}
+
+			private:
+				template<std::size_t... indicies>
+				void setAllCurrentIteratorsToTheirEnd(std::index_sequence<indicies...>){
+					auto setCurrentIteratorsToEnd = [](auto& iterTuple) {
+						std::get<1>(iterTuple) = std::get<2>(iterTuple);
+					};
+					(setCurrentIteratorsToEnd(std::get<indicies>(this->m_beginCurrenEndIterTuples)), ...);
+				}
+
+				template<std::size_t... indicies>
+				decltype(auto) makeTupleOfIteratorValues(std::index_sequence<indicies...>)const {
+					return std::make_tuple((*std::get<1>(std::get<indicies>(this->m_beginCurrenEndIterTuples)))...);
+				}
+
+				template<std::size_t... indicies>
+				void advanceCartezianIterator(std::index_sequence<indicies...>) {
+					makeCartezianIteratorPointToNextElement(std::get<indicies>(this->m_beginCurrenEndIterTuples)...);
+				}
+
+				static bool makeCartezianIteratorPointToNextElement() { return false; }
+
+				template<typename FirstIterator, typename... OtherIterators>
+				static bool makeCartezianIteratorPointToNextElement(
+					std::tuple<FirstIterator, FirstIterator, FirstIterator>& firstIterTuple,
+					std::tuple<OtherIterators, OtherIterators, OtherIterators>&... otherIterTuples) {
+
+					//if before incrementing the current iterator == end iterator
+					//then it means the whole iterator points to the end
+					if (std::get<1>(firstIterTuple) == std::get<2>(firstIterTuple))
+						return false;
+					++std::get<1>(firstIterTuple);
+					//if afer incrementig the current iterator == end iterator
+					//then we need to increment the first iterator in otherIterators recursively
+					if (std::get<1>(firstIterTuple) == std::get<2>(firstIterTuple)) {
+						if (makeCartezianIteratorPointToNextElement(otherIterTuples...)) {
+							std::get<1>(firstIterTuple) = std::get<0>(firstIterTuple);
+							return true;
+						}
+						else
+							return false;
+					}
+					return true;
+				}
```

Comment:
consider adding comments as to where each namespace ends.

you could maybe read Google Coding Standard. While I don't agree with everything they say - they have good suggestions and practices!